lib/runners/processor/clang_tidy.rb

```diff
@@ -0,0 +1,106 @@
+  class Processor::ClangTidy < Processor
+    Schema = StrongJSON.new do
+      let :runner_config, Schema::BaseConfig.base.update_fields { |fields|
+        fields.merge!({
+          apt: enum?(string, array(string)),
+          'compilation-options': object?(
+            '-I': enum?(string, array(string))
+          )
+        })
+      }
+
+      let :issue, object(
+        severity: string,
+      )
+    end
+
+    register_config_schema(name: :clang_tidy, schema: Schema.runner_config)
+
+    VALID_EXTENSIONS = [".c", ".cc", ".cpp", ".c++", ".cp", ".cxx"].freeze
+    GLOB_HEADERS = "**/*.{h,hpp,h++}".freeze
+
+    def analyze(changes)
+      deploy_packages
+      run_analyzer(changes)
+    end
+
+    private
+
+    def deploy_packages
+      # select "lib*" and report others as warning for security concerns
+      packages = Array(config_linter[:apt])
+        .select do |pkg|
+          if pkg.start_with?("lib")
+            true
+          else
+            add_warning "Installing the package '%s' is blocked." % pkg
+            false
+          end
+        end
+
+      unless packages.empty?
+        capture3!("sudo", "apt-get", "install", "-y", "--no-install-recommends", *packages)
+      end
+    end
+
+    def analyzer_bin
+      "clang-tidy-10"
+    end
+
+    def run_analyzer(changes)
+      issues = []
+
+      changes
+        .changed_paths
+        .select { |path| VALID_EXTENSIONS.include?(path.extname.downcase) }
+        .map{ |path| relative_path(working_dir / path, from: current_dir) }
+        .each do |path|
+          stdout, = capture3!(analyzer_bin, path.to_s, "--", *option_includes,
+            is_success: ->(status) { [0, 1].include?(status.exitstatus) })
+          ret = construct_result(stdout)
+          issues.push(*ret)
+        end
+
+      Results::Success.new(guid: guid, analyzer: analyzer).tap do |result|
+        result.add_issue(*issues)
+      end
+    end
+
+    def construct_result(stdout)
+      issues = []
+
+      stdout.each_line do |line|
+        # issue format
+        # <path>:<line>:<column>: <severity>: <message> [<id>]
+        match = line.strip.match(/^(?<path>.+):(?<line>\d+):(?<column>\d+): (?<severity>[^:]+): (?<message>.+) \[(?<id>[^\[]+)\]$/)
+        if match
+          issues << Issue.new(
+            path: relative_path(match[:path]),
+            location: Location.new(start_line: Integer(match[:line]), start_column: Integer(match[:column])),
+            id: match[:id],
+            message: match[:message],
+            object: {
+              severity: match[:severity],
+            },
+            schema: Schema.issue,
+          )
+        end
+      end
+
+      issues
+    end
+
+    def option_includes
+      includes = config_linter.dig(:'compilation-options', :'-I') || find_paths_containing_headers
+      includes.map { |v| "-I" + v }
+    end
+
+    def find_paths_containing_headers
+      working_dir.glob(GLOB_HEADERS, File::FNM_EXTGLOB | File::FNM_CASEFOLD)
```

Comment:
[suggestion] `FNM_CASEFOLD` seems unavailable for `Dir.glob`:

> Makes File.fnmatch patterns case insensitive (but not Dir.glob patterns).

```suggestion
      working_dir.glob(GLOB_HEADERS, File::FNM_EXTGLOB)
```

https://ruby-doc.org/core/File/Constants.html#FNM_CASEFOLD

Reply:
The clang-tidy's `-I` options should be relative from the working path where the `clang-tidy` command has been executed.

The `working_dir` variable is an instance of `Pathname` class which accepts `File::FNM_CASEFOLD` as bellow.
Indeed, the smoke test named `option_I_default` covers a test for the capital letters.

```shell
kentaro@debian103amd64:~/depot/runners/test/smokes/clang_tidy/success$ tree
.
+-- foo
|   +-- テスト.cpp
+-- myinclude
|   +-- テスト.hpp
+-- test.c

2 directories, 3 files
kentaro@debian103amd64:~/depot/runners/test/smokes/clang_tidy/success$ ruby -rpathname -e 'p Pathname(".").glob("**/*.HpP", File::FNM_EXTGLOB | File::FNM_CASEFOLD)'
[#<Pathname:myinclude/テスト.hpp>]
kentaro@debian103amd64:~/depot/runners/test/smokes/clang_tidy/success$ ruby -rpathname -e 'p Pathname(".").glob("**/*.HpP", File::FNM_EXTGLOB)'
[]
kentaro@debian103amd64:~/depot/runners/test/smokes/clang_tidy/success$
```

Reply:
Could it be that the `current_dir` is where the shell command is executed?
I'm not sure exact difference between `current_dir` and `working_dir`.

Reply:
Using `File::FNM_CASEFOLD` seems good. 👍 

---

> I'm not sure exact difference between `current_dir` and `working_dir`.

`current_dir` is a location which the `root_dir` option points at:

https://github.com/sider/runners/blob/87985b85b13fec23274aa719c2a3c5231dd9e10b/lib/runners/harness.rb#L47

https://github.com/sider/runners/blob/87985b85b13fec23274aa719c2a3c5231dd9e10b/lib/runners/processor.rb#L126-L132

https://github.com/sider/runners/blob/87985b85b13fec23274aa719c2a3c5231dd9e10b/lib/runners/shell.rb#L50

On the other hand, `working_dir` is the root of a repository.

Reply:
Thank you for making it clear. `current_dir` is suitable.
commit 027d4c5e0ba8c1f520f17c3a67af4b825078e6ef